src/lib/middleware/pageCacheMiddleware.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  // @ts-ignore
+  PAGE_CACHE_ENABLED,
+  // @ts-ignore
+  PAGE_CACHE_TYPES,
+  // @ts-ignore
+  PAGE_CACHE_NAMESPACE,
+  // @ts-ignore
+  PAGE_CACHE_VERSION,
+  // @ts-ignore
+  PAGE_CACHE_TIMEOUT_MINUTES,
+} from "../../config"
+import { Request, Response, NextFunction } from "express"
+const cache = require("lib/cache.coffee")
+const cacheablePageTypes: string[] = PAGE_CACHE_TYPES.split("|")
+
+// Middleware will `next` and do nothing if any of the following is true:
+//
+// * page cache feature is disabled.
+// * a user is signed in.
+// * this isnt a supported cacheable path (there is an allow-list set in ENV).
+//
+// At this point, the HTML is looked up in the cache.
+// If available, the request can be responded to and ended immediately.
+// Otherwise, `next` is called and the request continues in our app.
+export const pageCacheMiddleware = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  if (!PAGE_CACHE_ENABLED) return next()
+
+  const pageType = req.path.split("/")[1]
+  if (!cacheablePath(pageType)) return next()
+
+  // @ts-ignore
+  const hasUser = !!req.user
+  if (hasUser) return next()
+
+  const handleCacheGet = (err, html) => {
+    if (err) return next()
+
+    if (html) return res.send(html)
+
+    return next()
+  }
+
+  cache.get(cacheKey(req.url), handleCacheGet)
```

Comment:
In the past (in Metaphysics, I think), we've had surprising failures related to cache operations timing out or [worse] the cache being unavailable but failing to time out. Does this operation time out, and after how long? (A pretty low time-out would be appropriate, since the page should still be able to render without the cache.)

Reply:
Good call, this blocks b/c `next()` is only ever called in `handleCacheGet`, which as you point out, might be unexpectedly delayed if the cache is experiencing issues. Would be good to have that low time-out (as we do in Metaphysics for its cache).